src/happy/oauth2_capture_redirect.clj

```diff
@@ -14,19 +14,67 @@
       nil)))
 
-(defn wait-for-redirect [{:as config :keys [redirect_uri redirect_uris]}]
+(defn redirect-uri->port
+  "Opening a browser and redirecting to a local server
+  only makes sense for localhost redirect-uris.
+
+  Known limitation: Google does not support no ipv6 [::1] with random ports"
+  [uri]
+  (let [[match protocol host _ port path] (re-find #"^(http://)(localhost|127.0.0.1)(:(\d+))?(/.*)?$" uri)
+        port (maybe-parse-int port)]
+    (if (not match)
+      (throw (RuntimeException. "URI doesn't match pattern"))
+      (or port
+          0)))) ;; Jetty's magic value 0 = random port
+
+(defn add-port
+  [uri random-port]
+  (let [[match protocol host _ port path] (re-find #"^(http://)(localhost|127.0.0.1)(:(\d+))?(/.*)?$" uri)]
+    (if (not match)
+      (throw (RuntimeException. "URI doesn't match pattern"))
+      (str protocol host ":" random-port path))))
+
+(defn query-string->params
+  "Converts an URI query-string a=b&c=d&e=f&...
+  into a param map {:a \"b\" :c \"d\" :e \"f\"}."
+  [q]
+  (let [split-by-& #(str/split % #"&")
+        split-by-= #(str/split % #"=")
+        pairs (map split-by-= (split-by-& q))
+        parameters (into {} (for [[k v] pairs]
+                              [(keyword k) v]))]
+    parameters))
+
+(defn wait-for-redirect
+  "wait-for-redirect takes a callback function and a Google client
+  config. The callback function will be called, when the jetty server
+  has been started. It receives a single argument with the effective
+  URI, i.e. the URI with the correct (maybe random) port."
+  [on-server-started-callback
+   {:as config :keys [redirect_uri redirect_uris]}]
   (let [code (promise)
-        port (or (some-> (or redirect_uri (last redirect_uris))
-                         (str/split #":")
-                         (last)
-                         (maybe-parse-int))
-                 80)
-        server (jetty/run-jetty
-                 (fn [request]
-                   (deliver code (some-> request :query-string (str/split #"=") second))
-                   {:status 200
-                    :body   "ok"})
-                 (merge {:port port :join? false} config))]
+        uri (or redirect_uri (last redirect_uris))
+        port (redirect-uri->port uri)
+        http-redirect-handler (fn [request]
+                                (deliver code (some-> request
+                                                      :query-string
+                                                      query-string->params
+                                                      :code))
+                                {:status 200
+                                 :body   "ok"})
+        server (jetty/run-jetty http-redirect-handler
+                                {:port port :join? false})
+        _ (.setStopTimeout server 1000) ;; wait 1000ms for graceful shutdown, to avoid races between stopping and returning the response
+        
+        ^org.eclipse.jetty.server.NetworkConnector
```

Comment:
I added a type-hint here to allow the Clojure compiler to ahead-of-time compile and avoid reflection.